Expose hypergeometric_2F1 function

andrjohns opened the pull request:
## Summary

This PR adds the `hypergeometric_2F1` function, using the existing `grad_2f1` function to calculate the gradients.

## Tests

Minimal `mix` tests are added as the values are calculated using `boost::math::hypergeometric_pFq`, which is tested through the `hypergeometric_pFq` tests, and the gradients through `grad_2F1` which already has extensive tests. 

## Side Effects

N/A

## Release notes

Added `hypergeometric_2F1` function

## Checklist

- [x] Math issue #2664

- [x] Copyright holder: Andrew Johnson

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

SteveBronder reviewed (changes requested):
Couple small comments (and one optional big one)

andrjohns commented on stan/math/prim/fun/hypergeometric_2F1.hpp:
```diff
@@ -0,0 +1,44 @@
+  check_2F1_converges("hypergeometric_2F1", a1, a2, b, z);
+
+  return boost::math::hypergeometric_pFq({a1, a2}, {b}, z);
+}
```
I actually have a rough plan for an arbitrary scalar vectorisation framework, with signature something like:
```
apply_scalar(f, args...);
```

Once the ternary vectorisation goes in, then I'll be working that up

andrjohns commented:
@SteveBronder just flagging that I'm still adding some special-case handling, so don't worry about re-reviewing just yet. Thanks!

andrjohns commented:
@spinkey I've added a bunch of closed-form specialisations/reductions of the `hypergeometric_2F1` from the [WolframAlpha reference docs](https://functions.wolfram.com/HypergeometricFunctions/Hypergeometric2F1/03/), and also added the [Euler Transformation](https://mathworld.wolfram.com/EulerTransform.html) if the arguments don't meet convergence criteria:

$$
_2F_1(a,b;c;z) = \frac{_2F_1(c-a,b;c;\frac{z}{z-1})}{(1-z)^b}
$$

Which gives a bit of a wider domain of inputs. Do you know if there are any other transformations/reductions/etc that would be good to have in?

SteveBronder commented on stan/math/prim/fun/grad_inc_beta.hpp:
```diff
@@ -31,8 +31,9 @@ inline void grad_inc_beta(double& g1, double& g2, double a, double b,
+  double dFz = 0;
   if (C) {
-    grad_2F1(dF1, dF2, dF3, a + b, 1.0, a + 1, z);
+    grad_2F1(dF1, dF2, dF3, dFz, a + b, 1.0, a + 1, z);
```
[optional] instead of in/out parameters you could do something like 

```suggestion
    std::forward_as_tuple(dF1, dF2, dF3, dFz) = grad_2F1(a + b, 1.0, a + 1, z);
```

https://godbolt.org/z/4GPqjeYYf

SteveBronder commented on stan/math/prim/fun/hypergeometric_2F1.hpp:
```diff
@@ -0,0 +1,207 @@
+    return special_case_a2a1.get();
+  }
+
+  Eigen::VectorXd a_args(2);
+  Eigen::VectorXd b_args(1);
```
Sized at compile time

SteveBronder commented on stan/math/prim/fun/hypergeometric_2F1.hpp:
```diff
@@ -0,0 +1,207 @@
+    a_args << a1, a2;
+    b_args << b;
+    return hypergeometric_pFq(a_args, b_args, z);
+  } catch (...) {
```
idt we normally use ..., are there several types of errors that can be thrown?  I think our usual is to catch `const std::exception&` and if we don't want to do anything with it we leave it unnamed

SteveBronder reviewed (approved):
Looks good! I have a few comments for things you can optionally resolve but overall I think its clean

SteveBronder reviewed (changes requested):
Couple of Qs around `grad_2F1_impl`. I think we should either return a tuple or do in/out parameters, but I think doing both just makes things look a bit hard to read. Also I'm a little confused on why you need the user defined template parameter bool to specify whether the return should based on the `partial_type_t` or the `return_type_t`

andrjohns commented:
@SteveBronder This is ready for another look, I've tidied up the tuple/multiple-assignment handling and split out the Euler transform flow, so hopefully the codes a little more readable now

andrjohns commented on test/unit/math/prim/prob/wishart_cholesky_test.cpp:
```diff
@@ -81,7 +81,7 @@ TEST(ProbDistributionsWishartCholesky, dof_0) {
   MatrixXd L_Y = Y.llt().matrixL();
   MatrixXd L_S = Sigma.llt().matrixL();
 
-  unsigned int dof = std::numeric_limits<double>::quiet_NaN();
```
This change was needed to fix the `wishart_cholesky` prim failure from [this earlier Jenkins run](https://jenkins.flatironinstitute.org/blue/organizations/jenkins/Stan%2FMath/detail/PR-2792/21/pipeline/198)

This was relying on undefined behavior of a `NaN` overflowing to 0 when assigned to an int, which looks like it didn't happen in that Jenkins run. Let me know if you'd prefer I open separate PR and issue for this

SteveBronder commented:
Thanks!! That's fine and I'll look at this on Monday

SteveBronder commented on stan/math/prim/fun/grad_2F1.hpp:
```diff
@@ -9,144 +9,327 @@
+ * @param[in] max_steps number of steps to take
+ */
+template <bool ReturnSameT, typename T1, typename T2, typename T3, typename T_z,
+          require_not_t<std::integral_constant<bool, ReturnSameT>>* = nullptr>
```
[optional] 

```suggestion
          require_not_t<bool_constant<ReturnSameT>>* = nullptr>
```

SteveBronder commented on stan/math/prim/fun/grad_2F1.hpp:
```diff
@@ -9,144 +9,327 @@
+ *   to truncate the sum
+ * @param[in] max_steps number of steps to take
+ */
+template <bool ReturnSameT, typename T1, typename T2, typename T3, typename T_z,
+          require_t<std::integral_constant<bool, ReturnSameT>>* = nullptr>
+auto grad_2F1(const T1& a1, const T2& a2, const T3& b1, const T_z& z,
+              double precision = 1e-14, int max_steps = 1e6) {
+  return internal::grad_2F1_impl<true, true, true, true>(a1, a2, b1, z,
+                                                         precision, max_steps);
+}
```
If you set a default value for `ReturnSameT` like `bool ReturnSameT = false` do you need the other specialization below this one?

andrjohns replied:
Possibly, but since the default values have to come last in the template list I'd have to make some changes to how function would get called. I might leave that as a future TODO

SteveBronder reviewed (approved):
Two small optionals but looks good!
